Call wallet notify callbacks in scheduler thread (without cs_main)

TheBlueMatt opened the pull request:
Based on #10179, this effectively reverts #9583, regaining most of the original speedups of #7946.

This concludes the work of #9725, #10178, and #10179.

See individual commit messages for more information.

TheBlueMatt committed:
Add ability to assert a lock is not held in DEBUG_LOCKORDER

TheBlueMatt committed:
Add CWallet::BlockUntilSyncedToCurrentChain()

This blocks until the wallet has synced up to the current height.

TheBlueMatt committed:
Add a CValidationInterface::TransactionRemovedFromMempool

This is currently unused, but will by used by wallet to cache when
transactions are in the mempool, obviating the need for calls to
mempool from CWalletTx::InMempool()

TheBlueMatt committed:
Use callbacks to cache whether wallet transactions are in mempool

This avoid calling out to mempool state during coin selection,
balance calculation, etc. In the next commit we ensure all wallet
callbacks from CValidationInterface happen in the same queue,
serialized with each other. This helps to avoid re-introducing one
of the issues described in #9584 [1] by further disconnecting
wallet from current chain/mempool state.

Thanks to @morcos for the suggestion to do this.

Note that there are several race conditions introduced here:

 * If a user calls sendrawtransaction from RPC, adding a
   transaction which is "trusted" (ie from them) and pays them
   change, it may not be immediately used by coin selection until
   the notification callbacks finish running. No such race is
   introduced in normal transaction-sending RPCs as this case is
   explicitly handled.

 * Until Block{Connected,Disconnected} and
   TransactionAddedToMempool calls also run in the CSceduler
   background thread, there is a race where
   TransactionAddedToMempool might be called after a
   Block{Connected,Disconnected} call happens.

 * Wallet will write a new best chain from the SetBestChain
   callback prior to having processed the transaction from that
   block.

[1] "you could go to select coins, need to use 0-conf change, but
such 0-conf change may have been included in a block who's
callbacks have not yet been processed - resulting in thinking they
are not in mempool and, thus, not selectable."

ryanofsky reviewed (commented):
utACK 6fb571977d9cc41793a594688e5071dd5bbd864d.

Code changes mostly seem great, though as you can tell from my comments I have a somewhat hazy understanding of the semantics and assumptions being made. A little more documentation would make everything clear, I think.

TheBlueMatt commented on src/txmempool.h:
```diff
@@ -511,6 +511,9 @@ class CTxMemPool
     // then invoke the second version.
+    // Note that addUnchecked is ONLY called from ATMP during normal operation,
+    // and any other callers may break wallet's in-mempool tracking (due to
+    // lack of CValidationInterface::TransactionAddedToMempool callbacks).
```
Updated the comment to mention that addUnchecked is only called from ATMP outside of tests period. I think the implication is that we need to fix the strong-coupling here.

TheBlueMatt commented:
Rebased and fixed @ryanofsky's mostly-comment nits :).

TheBlueMatt committed:
Add a dev notes document describing the new wallet RPC blocking

mchrostowski reviewed (changes requested):
Overall appears to be on a good track. It looks to me like the global lock (cs_main) is causing some serious confusion/issues and there is some general mistake in the pattern of locks or their encapsulation that makes this all difficult. 

I reviewed everything pretty closely aside from the ZMQ test changes, that went over my head. 

Please don't overlook the outdated validationinterface.cpp comments, those took some time to put together.

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
+    void MaybeScheduleProcessQueue() {
+        {
+            LOCK(cs_callbacksPending);
+            // Try to avoid scheduling too many copies here, but if we
```
> This comment and issue can be avoided entirely if you move line 56 up to 46.
> After that lines 54 and 55 (which will be 55 and 56) can be removed.

I think this is right (line numbers apply to commit 8daf2439796dfdee41c1a32787e0ec9726daf6be). It also seems like you could eliminate the fCallbacksRunning variable if you change ProcessQueue to call pop_front after running the callback and condition the AddToProcessQueue schedule() call on the queue being previously empty.

TheBlueMatt replied:
Looks like you commented on an outdated version and github wont show me full context, so I have no idea what those line numbers refer to :/

mchrostowski replied:
@ryanofsky I believe the point is to avoid duplicate calls to the scheduler since it may be multi-threaded. So a call to AddToProcessQueue should not schedule() anything if we're already scheduled; it should only schedule() if our previously scheduled function has completed execution (at least beyond the point of it calling schedule() again). Can't see implementing that without knowing if fCallbacksRunning is true.

Still, the code does not guarantee single threaded execution in its current form due to the fCallbacksRunning state being set AFTER the call to schedule(). Being that all the locks are placed appropriately we might as well make this guarantee or else it seems like a bug because the SingleThreadedClient won't be.

I'll see if I can't get a test showing this behavior.

ryanofsky replied:
> @ryanofsky I believe the point is to avoid duplicate calls to the scheduler since it may be multi-threaded.

I know, this is why the second half of my suggestion was "condition the AddToProcessQueue schedule() call on the queue being previously empty." Anyway, I don't think Matt's particularly interested in these simplifications, and it's easier to communicate these changes as patches rather than english descriptions, so I'd rather just leave any simplifications to followup PRs.

> Still, the code does not guarantee single threaded execution in its current form due to the fCallbacksRunning state being set AFTER the call to schedule()

The reason it works in it's current form is because of the `if (fCallbacksRunning) return;` line at the top of `ProcessQueue()`

Again, I don't think the code in it's current form is the simplest it could be, but it seems safe and easy to clean up later in a followup PR. Also this whole discussion really should be moved to #10179. #10286 is only building on the changes in #10179.

mchrostowski replied:
@ryanofsky I see that now, the extra check does prevent the execution. 

Being that this is new code I wouldn't call it a simplification. Here's a patch of the proposed change, less logic with the same function:
[scheduler.patch.txt](https://github.com/bitcoin/bitcoin/files/977512/scheduler.patch.txt)

which reads better if you rename `fCallbacksRunning` to `fCallbacksScheduled`
and this patch:
[scheduler.patch2.txt](https://github.com/bitcoin/bitcoin/files/977548/scheduler.patch2.txt)
which can be argued reduces code reuse but I think the readability is improved.

mchrostowski replied:
@TheBlueMatt If you're open to these changes in a PR to your branch I can do that, I assume they'll be squashed so either way works.

TheBlueMatt replied:
@mchrostowski hmm, really, I find that it decreases readability (though that may be NIH). It looks harder to reason about whether some callbacks might accidentally get missed to me.

(Other random note, we dont use tabs in our codebase, which your patch added).

mchrostowski replied:
@TheBlueMatt Well, in that case I feel like either patch gets funky, especially since the use of fCallbacksRunning becomes inconsistent if you apply the first patch without the second (unless some alternative name for `fCallbacksRunning` works). 

The extra safety check and inconsistency of scheduling bothers me but I wouldn't expect it to actually cause issues so I have no grounds for objection.

I think my inquiry stemmed from it not being immediately apparent that ProcessQueue() only runs once and the extra check is just an extra check. Perhaps the "not a big deal" part of the comment could be "because ProcessQueue() already checks" for clarity, I would not have looked so deeply into the code except that I thought "not a big deal" meant "sometimes interweaving calls is okay."

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
+            callbacksPending.pop_front();
+        }
+
+        // RAII the setting of fCallbacksRunning and calling MaybeScheduleProcessQueue
```
> Why not try{} catch{}?

My guess about this was that it allows the processqueue to take advantage of whatever error handling or reporting cscheduler provides, and to not have to repeat the finalization logic both inside and after the catch clause. Either approach seems fine to me, though.

TheBlueMatt replied:
@laanwj previously requested that any exceptions be thrown all the way up, so this was an easier way to do that. That request seemed reasonable.

mchrostowski replied:
Understood, makes perfect sense. Didn't occur to me we'd have to duplicate the logic, spoiled by finally.

TheBlueMatt commented on src/qt/test/rpcnestedtests.cpp:
```diff
@@ -12,6 +12,7 @@
 #include "rpc/server.h"
```
Without them test_bitcoin-qt segfaults.

TheBlueMatt commented on src/validationinterface.cpp:
```diff
@@ -104,7 +104,9 @@ void UnregisterAllValidationInterfaces() {
 void CMainSignals::MempoolEntryRemoved(CTransactionRef ptx, MemPoolRemovalReason reason) {
     if (reason != MemPoolRemovalReason::BLOCK && reason != MemPoolRemovalReason::CONFLICT) {
-        internals->TransactionRemovedFromMempool(ptx);
+        internals->schedulerClient.AddToProcessQueue([ptx, this] {
```
This is being run in a background thread, so there are no possible lock inversions? We've had great success with DEBUG_LOCKORDER and havent had any serious deadlock issues afair since the 0.3.X era.

ryanofsky commented on src/validationinterface.cpp:
```diff
@@ -22,6 +27,64 @@ struct CMainSignalsInstance {
+                    LOCK(instance->cs_callbacksPending);
+                    instance->fCallbacksRunning = false;
+                }
+                instance->MaybeScheduleProcessQueue();
```
> Is this really what we want? A scheduler call for each callback?

I'm guessing the work done in any of these callbacks far outweighs the cost of scheduling one of them but I could be wrong.

TheBlueMatt replied:
Right now we only have one scheduler thread, so there should be limited lock contention, not to mention anything running is gonna take much longer than any locks held (locks are only held to push/pop/whatever, which should be very, very fast).

mchrostowski replied:
Okay, sounds like blocking/starving is the biggest risk. Can't disagree with that.

Feels like a bit of wasted work to call schedule on things you know are intended to execute now, but that's just a performance (not important) concern.
A bit surprising there is only one scheduler thread when the scheduler counts threads, but indeed it is only one.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1147,6 +1153,50 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
+        initialChainTip = chainActive.Tip();
+    }
+    AssertLockNotHeld(cs_main);
+    AssertLockNotHeld(cs_wallet);
```
The AssertLockNotHeld call only fails if the current thread holds the lock, not any thread.

mchrostowski replied:
`boost::thread_specific_ptr<LockStack> lockstack;`
Didn't see that, makes perfect sense then.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1147,6 +1153,50 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
 
 
 
+void CWallet::BlockUntilSyncedToCurrentChain() {
```
I'm not sure how it is not thread safe? It blocks the *current* thread, not any other work.

mchrostowski replied:
I thought unsafe due to the "AssertLockNotHeld" that I misunderstood. My above comment is totally wrong and can be disregarded.

TheBlueMatt commented on src/wallet/wallet.h:
```diff
@@ -1104,6 +1113,14 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
+    /**
+     * Blocks until the wallet state is up-to-date to /at least/ the current
+     * chain at the time this function is entered
+     * Obviously holding cs_main/cs_wallet when going into this call may cause
```
Any suggestions? The statement indicates that "holding cs_main/cs_wallet may cause deadlock", this is true, deadlock is not guaranteed, but may appear, thus you should obviously never call with cs_main or cs_wallet held.

mchrostowski replied:
I'd just drop "Obviously" now that I'm more familiar with the method.

TheBlueMatt commented on src/validationinterface.cpp:
```diff
@@ -5,48 +5,107 @@
+
 static CMainSignals g_signals;
 
+CMainSignals::CMainSignals() {
```
Does that compile? CMainSignalsInstance() is not defined at that time, only declared.

mchrostowski replied:
Since it is only a declaration I believe it should have been fine. I see this code isn't present in the final commit so I suppose it doesn't matter either way.

TheBlueMatt replied:
Ahh, yes, since the rebase on the latest version of #10179 this code no longer exists, as the scheduler has to be passed into the creation of the internals object.

TheBlueMatt commented on src/scheduler.h:
```diff
@@ -43,6 +43,9 @@ class CScheduler
     // Call func at/after time t
     void schedule(Function f, boost::chrono::system_clock::time_point t);
```
Took this on #10179, will be here when I next rebase.

TheBlueMatt commented on src/scheduler.h:
```diff
@@ -82,4 +84,26 @@ class CScheduler
+ * to be executed on the same thread, but no two jobs will be executed
+ * at the same time.
+ */
+class CSingleThreadedSchedulerClient {
```
Its unlikely to be used elsewhere, but as it is more intimate with the CScheduler than the validation interface, it was abstracted out and put here. We should probably tweak up how it all works in a later PR (as we move off of the big boost threadGroup in init), but for now I'll leave it.

ryanofsky commented on doc/developer-notes.md:
```diff
@@ -568,3 +568,16 @@ A few guidelines for introducing and reviewing new RPC interfaces:
+
+- Wallet RPCs call BlockUntilSyncedToCurrentChain to maintain consistency with
+  `getblockchaininfo`'s state immediately prior to the call's execution. Wallet
+  RPCs who's behavior does *not* depend on the current chainstate may omit this
```
In commit "Add a dev notes document describing the new wallet RPC blocking"

s/who's/whose

ryanofsky commented on doc/developer-notes.md:
```diff
@@ -568,3 +568,16 @@ A few guidelines for introducing and reviewing new RPC interfaces:
+  - *Rationale*: In previous versions of Bitcoin Core, the wallet was always
+    in-sync with the chainstate (by virtue of them all being updated in the
+    same cs_main lock). In order to maintain the behavior that wallet RPCs
+    return restults as of at least the highest best-known-block an RPC
```
In commit "Add a dev notes document describing the new wallet RPC blocking"

s/restults/results, s/best-known-block/best-known block,/

mchrostowski commented:
So far looks good to me, I'm going to poke around wallet<->blockchain interaction so I can better understand the wallet.cpp changes you made before I comment further.

That said I feel like there is something fundamentally wrong with the interaction between CWallet and the blockchain (I don't even know where that code lives yet). This feels like a solution to current issues but I would hope a refactor prevents needing such frequently occurring checks and thread specific execution. 

That said, it would be interesting if a single "blockchain operations" thread is needed to get the code working in a reasonable manner. GUI systems tend to need such threads because of the nature of problem which is events coming from both ends of the system (rendering thread vs input thread) which would have to, in any logical design, invert lock ordering or else excessively block.

ryanofsky reviewed (commented):
utACK 943217460bc527c4003868415c264e4a77a6e55a.

Changes since previous ACK were rebasing, adding developer notes, tweaking some comments, adding a check to the zmq test

TheBlueMatt commented:
@mchrostowski thanks for the review! Generally, wallet and blockchain (essentially validation.cpp's stuff) have historically been pretty tightly coupled (updated all under the same cs_main lock). This PR is a step, however small, towards decoupling that a bit. Because the wallet still relies on "is it in our mempool?" as a proxy for "is this possibly going to get confirmed/is it spendable with the result making it into my mempool", the chainstate and mempool still need to be updated in-sync and the wallet notified of the updates in a single notification, which come in in the order they happened. Still, this is much better than the wallet actually querying the mempool/validation logic directly instead of tracking the stuff it cares about out of them.

TheBlueMatt commented:
Rebased on latest #10179, current master, and fixed @ryanofsky's english corrections.

ryanofsky commented:
utACK 2c306d7876fb57ff26d217f97415a79942094002. Changes since previous were some documentation tweaks and new block calls in "Add calls to CWallet::BlockUntilSyncedToCurrentChain()"

mchrostowski commented:
@TheBlueMatt The stated purpose of this PR is to reduce locking on cs_main so as to reducing code coupling. I see one change in this PR that actually deletes a `LOCK(cs_main)` which is in `CWallet::InMempool()`. This looks like a step in the right direction.

That said, the remaining changes seem to be all about getting the signals into a background thread. What does this gain us for decoupling?

If the purpose is to not hold cs_main from whatever call sites hit `GetMainSignals()` then I don't see a benefit in using new threads when we can release the lock before making the call. That is, running in another thread is not decoupling synchronization or interface dependencies. The goal of "move wallet updates out of cs_main into a background thread" seems unrelated to decoupling because "using a background thread" and "not holding cs_main" are not dependent on each other, at least in the cases I observed.

Of particular concern is the re-ordering of calls, which you're avoiding with the single threaded scheduler, but if this isn't required for decoupling then it is just added complexity and overhead.

ryanofsky commented:
This needs rebase due to a minor conflict in listunspent.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1147,6 +1153,50 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
+                // If the user called invalidateblock our wait here might block
+                // forever, so we check if we're ahead of the tip (a state
+                // which should otherwise never be exposed outside of validation)
+                return this->lastBlockProcessed->nChainWork > chainActive.Tip()->nChainWork;
```
In commit "Add CWallet::BlockUntilSyncedToCurrentChain()"

I don't understand why we would wait forever without this check. Does InvalidateBlock not trigger notifications that would lead to lastBlockProcessed being updated? And if it doesn't, shouldn't this just be fixed so the right notifications are sent?

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1147,6 +1153,50 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
+            // This should be exceedingly rare in regular usage, so potentially
+            // eating 100ms to retry this lock should be fine (not TRY_LOCKing
+            // here would be a lock inversion against lastBlockProcessedMutex)
+            TRY_LOCK(cs_main, mainLocked);
```
In commit "Add CWallet::BlockUntilSyncedToCurrentChain()"

Maybe consider dropping the try-lock and replacing it with `lastBlockProcessedMutex.unlock(); LOCK(cs_main);  lastBlockProcessedMutex.lock();`. Maybe this would be a little slower in the average case where this code runs (which is rare to begin with), but it would avoid the 100ms worst case, and make the code simpler because you could also replace the while loop and timeout with a plain `cv.wait(lock, pred)` call.

TheBlueMatt committed:
Call TransactionRemovedFromMempool in the CScheduler thread

This is both good practice (we want to move all such callbacks
into a background thread eventually) and prevents a lock inversion
when we go to use this in wallet (mempool.cs->cs_wallet and
cs_wallet->mempool.cs would otherwise both be used).

TheBlueMatt committed:
Also call other wallet notify callbacks in scheduler thread

This runs Block{Connected,Disconnected}, SetBestChain, Inventory,
and TransactionAddedToMempool on the background scheduler thread.

Of those, only BlockConnected is used outside of Wallet/ZMQ, and
is used only for orphan transaction removal in net_processing,
something which does not need to be synchronous with anything
else.

This partially reverts #9583, re-enabling some of the gains from
 #7946. This does not, however, re-enable the gains achieved by
repeatedly releasing cs_main between each transaction processed.

TheBlueMatt committed:
Add CallFunctionInQueue to wait on validation interface queue drain

TheBlueMatt commented:
Rebased and rewrote CWallet::BlockUntilSyncedToCurrentChain(). Instead of the complicated fallback logic, it now just tests if it is caught up, and if it is not, it puts a callback into the CValidationInterface queue and waits for it to trigger. I wanted to avoid having this function previously, but I ended up needing it for a different branch which moves more CValidationInterface callbacks to the background and the logic is so simple, that I went ahead with it.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1154,6 +1156,39 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
+        }
+    }
+
+    std::condition_variable callbacks_done_cv;
```
In commit "Add CWallet::BlockUntilSyncedToCurrentChain()"

I think all the mutex/cv/lambda/looping stuff below could be replaced by:

```
std::packaged_task<void()> nop([]() {});
CallFunctionInValidationInterfaceQueue(std::ref(nop));
nop.get_future().wait();
```

ryanofsky reviewed (commented):
utACK 91aad9fa3b33ae387145e4b84a14c5f61cbd2494. Changes since last review were rebase, style guide fixes, BlockUntilSyncedToCurrentChain rewrite.

ryanofsky reviewed (commented):
utACK ae0c83326bdae2b61c1bcffd122f78468aa2aefb. Same as previous except the last 3 review suggestions are now incorporated. There is a minor conflict now with master, but since this depends on #10179 anyway, there's probably no hurry to rebase.

ryanofsky commented on src/rpc/rawtransaction.cpp:
```diff
@@ -880,6 +880,15 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
+            // where a user might call sendrawtransaction with a transaction
+            // to/from their wallet, immediately call some wallet RPC, and get
+            // a stale result because callbacks have not yet been processed.
+            pwalletMain->TransactionAddedToMempool(tx);
```
In commit "Fix wallet RPC race by informing wallet of tx in sendrawtransaction"

Seems like this will result in the wallet getting two TransactionAddedToMempool notifications, which is fine but might be worth noting in the comment.

Also, not asking for this change, but would another way to do this without referencing the wallet here be to release cs_main and then wait for the other notification to be processed? (Maybe using CallFunctionInValidationInterfaceQueue like in BlockUntilSyncedToCurrentChain?)

TheBlueMatt replied:
Yes, much better to CallFunctionInValidatioInterface. Done.

TheBlueMatt committed:
Fix wallet RPC race by waiting for callbacks in sendrawtransaction

ryanofsky commented on src/sync.cpp:
```diff
@@ -156,6 +156,15 @@ void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine,
 
+void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, void* cs)
+{
+    BOOST_FOREACH (const PAIRTYPE(void*, CLockLocation) & i, *lockstack)
```
In commit "Add ability to assert a lock is not held in DEBUG_LOCKORDER"

Looks like there are travis errors compiling this code (undefined FOREACH/PAIRTYPE).

ryanofsky commented on src/rpc/rawtransaction.cpp:
```diff
@@ -881,6 +883,17 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
+            CallFunctionInValidationInterfaceQueue([&promise] {
+                promise.set_value();
+            });
+            promise.get_future().wait();
```
In commit "Fix wallet RPC race by waiting for callbacks in sendrawtransaction"

I think you might need to release cs_main before waiting for the promise, because the wallet handler in the notification thread will want to acquire it.

FWIW, the way I implemented this in my wallet ipc branch was to call `CallFunctionInValidationInterfaceQueue` the same way you are doing here, but release cs_main before the `PushInventory` calls, and wait for the promise after the `PushInventory` calls. I also updated the comment.

```diff
--- a/src/rpc/rawtransaction.cpp
+++ b/src/rpc/rawtransaction.cpp
@@ -29,6 +29,7 @@
 #include "wallet/wallet.h"
 #endif
 
+#include <future>
 #include <stdint.h>
 
 #include <univalue.h>
@@ -846,7 +847,6 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
             + HelpExampleRpc("sendrawtransaction", "\"signedhex\"")
         );
 
-    LOCK(cs_main);
     RPCTypeCheck(request.params, {UniValue::VSTR, UniValue::VBOOL});
 
     // parse hex string from parameter
@@ -860,6 +860,9 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
     if (request.params.size() > 1 && request.params[1].get_bool())
         nMaxRawTxFee = 0;
 
+    std::promise<void> sent_notifications;
+    {
+    LOCK(cs_main);
     CCoinsViewCache &view = *pcoinsTip;
     bool fHaveChain = false;
     for (size_t o = 0; !fHaveChain && o < tx->vout.size(); o++) {
@@ -881,15 +884,6 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
                 }
                 throw JSONRPCError(RPC_TRANSACTION_ERROR, state.GetRejectReason());
             }
-#ifdef ENABLE_WALLET
-        } else {
-            // If wallet is enabled, ensure that the wallet has been made aware
-            // of the new transaction prior to returning. This prevents a race
-            // where a user might call sendrawtransaction with a transaction
-            // to/from their wallet, immediately call some wallet RPC, and get
-            // a stale result because callbacks have not yet been processed.
-            pwalletMain->TransactionAddedToMempool(tx);
-#endif
         }
     } else if (fHaveChain) {
         throw JSONRPCError(RPC_TRANSACTION_ALREADY_IN_CHAIN, "transaction already in block chain");
@@ -897,11 +891,22 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
     if(!g_connman)
         throw JSONRPCError(RPC_CLIENT_P2P_DISABLED, "Error: Peer-to-peer functionality missing or disabled");
 
+    CallFunctionInValidationInterfaceQueue([&sent_notifications] { sent_notifications.set_value(); });
+    } // LOCK(cs_main)
+
     CInv inv(MSG_TX, hashTx);
     g_connman->ForEachNode([&inv](CNode* pnode)
     {
         pnode->PushInventory(inv);
     });
+
+    // Wait for any TransactionAddedToMempool notifications sent above to be
+    // processed by thw wallet. This prevents a race where a user might call
+    // sendrawtransaction with a transaction to/from their wallet, immediately
+    // call some wallet RPC, and get a stale result because callbacks have not
+    // yet been processed.
+    sent_notifications.get_future().wait();
+
     return hashTx.GetHex();
 }
```

TheBlueMatt replied:
Heh, beat you to it.

ryanofsky reviewed (commented):
utACK db1d712d5668496ca83947cf9d6c4a39d3a3af1f. Changes since last review: rebase, sendraw change, naming style updates, added std::function includes.

ryanofsky commented on src/rpc/rawtransaction.cpp:
```diff
@@ -893,6 +901,22 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
+        CallFunctionInValidationInterfaceQueue([&promise] {
+            promise.set_value();
+        });
+        promise.get_future().wait();
```
In commit "Fix wallet RPC race by waiting for callbacks in sendrawtransaction"

This seems right. Possible tweaks:

* `should_wait_on_validationinterface` seems like it is always true at this point, maybe the variable is not needed.
* Might be better to call `CallFunctionInValidationInterfaceQueue` before releasing cs_main, to avoid waiting for notifications that might be queued in the meantime that we don't care about.
* Might be cheaper / simpler just to hold on to the tx reference (remove std::move) instead of copying the hash to a temporary variable.
* Maybe release cs_main before the PushInventory loop.

ryanofsky reviewed (commented):
utACK e3617741430df7d21d807549041f7251ba5b91e5. Only change since last review was sendraw locking fix.

ryanofsky commented on src/rpc/rawtransaction.cpp:
```diff
@@ -881,19 +887,34 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
+
+    } // cs_main
+
+    promise.get_future().wait();
```
In commit "Fix wallet RPC race by waiting for callbacks in sendrawtransaction"

Might be more efficient to wait for the promise after the PushInventory calls so they aren't blocked waiting for wallets.

theuni replied:
Without net/connman, I'd expect this RPC to fail immediately anyway.

ryanofsky reviewed (commented):
utACK 68201ec021944b1216e321a572060a7fd5183cbe. Just AssertLockNotHeldInternal foreach fix and sendrawtransaction tweaks since last review.

TheBlueMatt commented on src/rpc/rawtransaction.cpp:
```diff
@@ -881,19 +887,34 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
+
+    } // cs_main
+
+    promise.get_future().wait();
```
I figured put it above the !g_connman check to make sure we block even if at some point in the future we support running without net/connman. Dont feel strongly either way, though

TheBlueMatt commented:
Rebased. Would be nice to get this in early in 16 to let it simmer on master.

ryanofsky reviewed (commented):
utACK 964f376b6a34294d328b634ef045b11079b0859d. No changes since last review, only rebase.

jnewbery commented:
I haven't fully reviewed yet, but @ryanofsky pointed me to this for `m_last_block_processed`. Should that be rewound in `DisconnectBlock()`?

ryanofsky commented:
I think it should be safe to update m_last_block_processed in disconnectblock. It just isn't needed for this pr, because m_last_block_processed is used here in a pretty limited way.

promag commented on src/wallet/wallet.h:
```diff
@@ -337,6 +335,7 @@ class CWalletTx : public CMerkleTx
     mutable CAmount nImmatureWatchCreditCached;
     mutable CAmount nAvailableWatchCreditCached;
     mutable CAmount nChangeCached;
+    mutable bool fInMempool;
```
Move up for better packing?

promag replied:
Instead of `fInMempool` for each wallet transaction, maybe a `std::set<WalletTx*>` with the mempool transactions takes less memory?

TheBlueMatt replied:
Moved it up. At some point someone can go through and mark all the bools as 1 bit, I'd think.

promag commented on src/rpc/misc.cpp:
```diff
@@ -75,6 +75,9 @@ UniValue getinfo(const JSONRPCRequest& request)
 #ifdef ENABLE_WALLET
     CWallet * const pwallet = GetWalletForJSONRPCRequest(request);
 
+    if (pwallet) {
```
Move this to `EnsureWalletIsAvailable`?

TheBlueMatt replied:
Not all functions which want wallet need to do this, though.

promag reviewed (commented):
Concept ACK.

TheBlueMatt commented on src/validationinterface.cpp:
```diff
@@ -85,10 +97,17 @@ void UnregisterAllValidationInterfaces() {
     g_signals.m_internals->NewPoWValidBlock.disconnect_all_slots();
 }
 
+void CMainSignals::MempoolEntryRemoved(CTransactionRef ptx, MemPoolRemovalReason reason) {
```
This isnt a new function, it would have to be changed in txmempool.h

TheBlueMatt commented on src/validationinterface.cpp:
```diff
@@ -85,10 +97,17 @@ void UnregisterAllValidationInterfaces() {
 }
 
+void CMainSignals::MempoolEntryRemoved(CTransactionRef ptx, MemPoolRemovalReason reason) {
+    if (reason != MemPoolRemovalReason::BLOCK && reason != MemPoolRemovalReason::CONFLICT) {
```
Because BLOCK and CONFLICT are passed out in BlockConnectedDisconnected (see comment in validationinterface.h which describes in what cases this is called).

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1167,6 +1170,29 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
+    }
+
+    std::promise<void> promise;
+    CallFunctionInValidationInterfaceQueue([&promise] {
```
I think you've been bitten by GitHub's obnoxious order-commits-by-date-instead-of-git-ordering bug?

TheBlueMatt commented:
Rebased.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -4365,7 +4383,26 @@ int CMerkleTx::GetBlocksToMaturity() const
+    return ret;
+}
+
+bool CalculateEstimateType(FeeEstimateMode mode, bool opt_in_rbf) {
```
In commit "Use callbacks to cache whether wallet transactions are in mempool"

Looks like you accidentally resurrected this CalculateEstimateType function in the rebase. Should remove.

TheBlueMatt replied:
Oops, fixed.

ryanofsky reviewed (commented):
utACK 9adaf5f2c5a28e338dcf99d873f0bd40000696e7. Only change sinces the last review were rebase and fInMempool move

TheBlueMatt committed:
Add calls to CWallet::BlockUntilSyncedToCurrentChain() in RPCs

This prevents the wallet-RPCs-return-stale-info issue from being
re-introduced when new-block callbacks no longer happen in the
block-connection cs_main lock

TheBlueMatt commented:
(Trivially) rebased.

danra commented on src/validationinterface.h:
```diff
@@ -7,6 +7,7 @@
 #define BITCOIN_VALIDATIONINTERFACE_H
 
 #include <memory>
+#include <functional>
```
Since you're adding system includes, might as well move them to their correct place, after the user includes.

danra commented on src/wallet/wallet.h:
```diff
@@ -724,6 +724,14 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
     std::unique_ptr<CWalletDBWrapper> dbw;
 
+    /**
+     * The following are used to keep track of how far behind the wallet is
```
are used->is used?

danra commented on src/wallet/wallet.cpp:
```diff
@@ -3869,6 +3895,12 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
             pindexRescan = FindForkInGlobalIndex(chainActive, locator);
     }
+
+    //We must set m_last_block_processed prior to registering the wallet as a validation interface
```
Missing space after `//`

danra commented on src/validationinterface.h:
```diff
@@ -31,6 +32,11 @@ void RegisterValidationInterface(CValidationInterface* pwalletIn);
+ * Pushes a function to callback onto the notification queue, guaranteeing any
+ * callbacks generated prior to now are finished when the function is called.
+ */
+void CallFunctionInValidationInterfaceQueue(std::function<void ()> func);
```
I think a better interface would be
`std::future<void> SyncWithValidationInterfaceQueue();`

Whose implementation sets up the delayed call to a `promise.set_value()`, instead of this being the responsibility of the caller as it is now.

The sync functionality is the only purpose, so rather than have an overly-generic interface which allows calling any function, it's better to have a more specific interface which simplifies the interface as well as the setup work on the caller side.

TheBlueMatt replied:
Hmm..I have a few more branches which use this function for other stuff (eg one that will set a "allowed to continue processing" boolean on a CNodeState then call WakeMessageHandler). I think most of the use-cases for now could almost do fine with a future-like return value if std::future allowed you to query whether the future had yet completed. Still, since the promise-create-future-wait pattern isnt too many LoC, I suppose I'll leave it for now instead of adding a bunch of machinery just to create a special future-like return type.

danra replied:
> I think most of the use-cases for now could almost do fine with a future-like return value if `std::future` allowed you to query whether the future had yet completed.
Wouldn't `std::future::wait_for()` with argument provide that functionality?

TheBlueMatt replied:
Yea, could pull some hack like wait_for(1ns), but that's pretty gross. I think the purpose is somewhat of both - really it shouldn't be to do much more than sync, but sync may require rather arbitrary things (like calling WakeMessageHandler to wake up another thread which is waiting on some other unrelated condition in the example above). For now probably best to just leave it IMO.

danra replied:
Ok.
btw I meant to write "with argument *zero*" above, so it's not hacky - it just checks if the `std::future` is done, without waiting.

danra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2769,6 +2857,10 @@ UniValue listunspent(const JSONRPCRequest& request)
 
+    // Make sure the results are valid at least up to the most recent block
+    // the user could have gotten from another RPC command prior to now
+    pwallet->BlockUntilSyncedToCurrentChain();
```
This should come after `assert(pwallet != nullptr);`

TheBlueMatt replied:
Removed that line instead (we dont have similar checks anywhere else and the case is already covered by EnsureWalletIsAvailable).

danra commented:
- `BlockConnectedDisconnected` in comments and commit messages should say `BlockConnected/Disconnected`, or Block[Connected|Disconnected]`, or something of the sort. The way it's written like now looks like there's an actual single function by that name.
- In commit message 975f071b49b642b069c5e1de8d0050b663ca7c1c, "CSceduler background threa" -> "CScheduler background thread"

danra commented on src/validationinterface.cpp:
```diff
@@ -119,23 +119,33 @@ void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInd
-    m_internals->SetBestChain(locator);
+    m_internals->m_schedulerClient.AddToProcessQueue([locator, this] {
+        m_internals->SetBestChain(locator);
+    });
```
I think this introduces a bug. `CMainSignals::SetBestChain()` might be called with a `CBlockLocator` object argument which goes out of scope and gets destructed before the queued `m_internals->SetBestChain()` is called, causing an invalid memory access. For example, this can happen in the `walletInstance->SetBestChain(chainActive.GetLocator());` call in wallet.cpp).

danra replied:
On second thought I'm probably wrong because locator is captured by the lambda by value, not reference.

danra commented on src/rpc/rawtransaction.cpp:
```diff
@@ -31,6 +32,7 @@
 #endif
 
 #include <stdint.h>
+#include <future>
```
sort

danra commented:
utACK, looks like a very nice improvement.

TheBlueMatt committed:
Give ZMQ consistent order with UpdatedBlockTip on scheduler thread

Note that UpdatedBlockTip is also used in net_processing to
announce new blocks to peers. As this may need additional review,
this change is included in its own commit.

TheBlueMatt committed:
Remove redundant pwallet nullptr check

TheBlueMatt commented:
Rebased and address @danra's comments.

jnewbery commented on src/validationinterface.cpp:
```diff
@@ -85,32 +97,57 @@ void UnregisterAllValidationInterfaces() {
+    g_signals.m_internals->m_schedulerClient.AddToProcessQueue(std::move(func));
+}
+
+void CMainSignals::MempoolEntryRemoved(CTransactionRef ptx, MemPoolRemovalReason reason) {
```
Is there a reason that you've named the validation interface callback `TransactionRemovedFromMempool()` but the signal is named `MempoolEntryRemoved`. For all other callbacks in the validation interface, the callback name matches the signal name.

TheBlueMatt replied:
For symmetry with TransactionAddedToMempool. I'd rather rename the MempoolEntryRemoved signal in CTxMempool isntead, but, really, I cant say I have hugely strong feelings. Considering I have a PR lined up after this one to rework some of the mempool parts of this interface, maybe lets leave that for the next PR?

jnewbery commented on src/wallet/wallet.h:
```diff
@@ -722,6 +725,14 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
+     * The following is used to keep track of how far behind the wallet is
+     * from the chain sync, and to allow clients to block on us being caught up
+     *
+     * Protected by cs_main
```
I don't understand why this is protected by cs_main. Can you explain?

TheBlueMatt replied:
The code in BlockUntilSyncedToCurrentChain accesses this in a place that's going to need cs_main anyway (for chainActive.Tip()), could add a cs_wallet and call it protected by cs_wallet, but that just means an extra lock in BlockUntilSyncedToCurrentChain.

TheBlueMatt replied:
I added comments to explain this.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1625,6 +1666,8 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool f
 
         fScanningWallet = false;
+
+        m_last_block_processed = chainActive.Tip();
```
This will be set incorrectly if the rescan was aborted.

TheBlueMatt replied:
I just removed this and set it to chainActive.Tip() in CreateWalletFromFile. It doesnt make sense to try to use this to block wallet until we're caught up with the current tip in reorg, it isnt a regression so we should just leave it. When/if we expose last_block_processed via RPC (and allow wallet to return data that is stale) we should change that, I suppose.

jnewbery replied:
Yes - I agree. `m_last_block_processed` would be a really nice value to expose in the `getwalletinfo` RPC.

I have a weak preference to make `m_last_block_processed` work as you'd expect (tracking the last block processed by the wallet), but this PR is already doing a lot, so I think it's also fine to keep this as is. Please update the comment in wallet.h to say that it's not actually tracking how far the wallet is behind chain sync and shouldn't be publicly exposed until that's changed.

TheBlueMatt replied:
OK, added more info to the wallet.h comment.

jnewbery commented on src/rpc/rawtransaction.cpp:
```diff
@@ -917,14 +919,19 @@ UniValue sendrawtransaction(const JSONRPCRequest& request)
+    CTransactionRef tx;
+    std::promise<void> promise;
+
+    { // cs_main scope
```
Can you move this cs_main scope further down (to avoid having to declare the `CTransactionRef` above and then recompute the `GetHash()` at the end of this function?

TheBlueMatt replied:
CTransactionRef->GetHash() is (virtually) free because we cache the hash in the CTransaction. Honestly we should probably just go the other way and remove the hashTx local. Additionally, if I'm correctly understanding your request, that would result in a deadlock as the validation interface queue may not be able to make progress as long as cs_main is held in the RPC thread. I've added a comment to validationinterface.h above CallFunctionInValidationInterfaceQueue to this effect.

TheBlueMatt replied:
Misunderstood your point, moved the scope *start* down.

jnewbery replied:
Yes, sorry - my comment was unclear.

Change looks good.

jnewbery reviewed (commented):
Lightly tested ac204f8f87bbc761a4e682c084838a6f47349bae. A few comments and questions inline.

TheBlueMatt commented on src/validationinterface.cpp:
```diff
@@ -85,32 +97,57 @@ void UnregisterAllValidationInterfaces() {
+
 void CMainSignals::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload) {
-    m_internals->UpdatedBlockTip(pindexNew, pindexFork, fInitialDownload);
+    m_internals->m_schedulerClient.AddToProcessQueue([pindexNew, pindexFork, fInitialDownload, this] {
```
I believe this to be fine. nTimeBestReceived is passed to wallet as the first parameter to ResendWalletTransactions, and is effectively just called on a loop to resend wallet transactions when we need to. If it gets called with an nTimeBestReceived that is further back than we are, that should be fine, it just won't rebroadcast until the next time its called.

jnewbery replied:
Yep, sounds good. I *thought* this was ok, but I wasn't certain and wanted to check.

TheBlueMatt commented:
Addressed @jnewbery's comments, and added more code comments where things were not entirely clear.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1244,6 +1262,29 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
+    }
+
+    std::promise<void> promise;
+    CallFunctionInValidationInterfaceQueue([&promise] {
```
New comment looks good. Thanks.

jnewbery commented:
Most recent round of changes looks good to me. utACK 024e7b51ee4eeaf21ecda71915438241dd666705

TheBlueMatt commented:
Rebased.

laanwj commented:
utACK 89f0312

theuni commented on src/wallet/wallet.cpp:
```diff
@@ -1244,6 +1247,36 @@ void CWallet::BlockDisconnected(const std::shared_ptr<const CBlock>& pblock) {
+    CallFunctionInValidationInterfaceQueue([&promise] {
+        promise.set_value();
+    });
+    promise.get_future().wait();
```
Isn't the scheduler queue abandoned at shutdown? What keeps this from blocking forever if timed just right?

TheBlueMatt replied:
No, init.cpp calls CMainSignals().FlushBackgroundCallbacks() which will empty the CValidationInterfaceQueue. The scheduler itself is abandoned, but we shouldn't generate any new notifications by that point.

theuni commented on src/wallet/wallet.cpp:
```diff
@@ -3913,6 +3944,10 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
             pindexRescan = FindForkInGlobalIndex(chainActive, locator);
     }
+
+    walletInstance->m_last_block_processed = chainActive.Tip();
```
Doesn't this entire function need cs_main ?

TheBlueMatt replied:
It would appear so (though that isn't introduced here)....I'll let @practicalswift take a look since that appears to be his project atm.

practicalswift replied:
Yes, according to my annotations (#11226) calling `CreateWalletFromFile` requires holding `cs_main`:

```
src/wallet/wallet.h:    static CWallet* CreateWalletFromFile(const std::string walletFile)
                            EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```

This is due to:
* `CreateWalletFromFile` is reading the variable `chainActive` which requires holding the mutex `cs_main`.
* `CreateWalletFromFile` calls `FindForkInGlobalIndex` which reads the variable `mapBlockIndex`. Reading the variable `mapBlockIndex` requires holding the mutex `cs_main`.

theuni commented on src/wallet/wallet.cpp:
```diff
@@ -3010,14 +3024,18 @@ bool CWallet::CommitTransaction(CWalletTx& wtxNew, CReserveKey& reservekey, CCon
 
+        // Get the inserted-CWalletTx from mapWallet so that the
+        // fInMempool flag is cached properly
+        CWalletTx& wtx = mapWallet[wtxNew.GetHash()];
```
Somewhat unrelated nit: AddToWallet's return value should be checked before using it here, but that was already the case.

TheBlueMatt replied:
Indeed, though in such cases not much we can do - bdb got upset, we're probably just going to shutdown soon anyway.

morcos commented on src/init.cpp:
```diff
@@ -265,6 +265,7 @@ void Shutdown()
 #endif
     UnregisterAllValidationInterfaces();
     GetMainSignals().UnregisterBackgroundSignalScheduler();
+    GetMainSignals().UnregisterWithMempoolSignals(mempool);
```
@TheBlueMatt Shouldn't this line be before the `UnregisterBackgroundSignalScheduler`?

TheBlueMatt replied:
I suppose it would be more correct, yes, though I do not believe this is a bug - mempool should absolutely, absolutely not be generating any events by this point, so if it does (which I believe would result in an assert(false) or so) it would represent some other shutdown-order bug.

ryanofsky commented:
According to https://github.com/bitcoin/bitcoin/issues/11822#issuecomment-348903894, this may be causing a memory leak
